### PR TITLE
[Android/tflite] Fix NNAPI inference bug

### LINKED ISSUE #2755: [Android/tflite] Failed to invoke tflite model with nnapi.
The test `testInvokeDynamicTFLite()` keeps failing by crash.
(https://github.com/nnstreamer/nnstreamer/blob/for-next/api/android/api/src/androidTest/java/org/nnsuite/nnstreamer/APITestSingleShot.java#L272)

Currently, tflite (1.13.1) utilizes nnapi by default: `use_nnapi = TRUE;`
(https://github.com/nnstreamer/nnstreamer/blob/master/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc#L574)

If I modify this line with `use_nnapi = FALSE;`, the failures do not occur.
This may be related with file access permission.

Test failure log:
```
...
2020-09-21 16:03:25.046 4220-4323/? D/nnstreamer: Framework = tensorflow-lite
2020-09-21 16:03:25.046 4220-4323/? I/GLib: nnapi = 1, accl = cpu
2020-09-21 16:03:25.047 684-684/? D/keymaster_swd: keymaster_swd [WRN] (swd_run_cb:249) swd_finish() returns 0
2020-09-21 16:03:25.047 4220-4323/? W/nnstreamer: The number of tensor name is mismatched in filter.
2020-09-21 16:03:25.047 4220-4323/? W/nnstreamer: The number of tensor name is mismatched in filter.
2020-09-21 16:03:25.047 1053-4566/? D/keystore: [FINISH::end]
2020-09-21 16:03:25.048 4220-4323/? E/NeuralNetworks: SecFeature = 3
2020-09-21 16:03:25.050 4220-4323/? D/NativeHcmServiceManager: [NativeHcm] BpSemHcmManager::BpSemHcmManager()
2020-09-21 16:03:25.050 1055-12636/? I/HcmInfo::HcmManagerService: ctc pid=4220, uid=10131, Package Name:org.nnsuite.nnstreamer.test
2020-09-21 16:03:25.051 1055-2553/? D/CustomFrequencyManagerService: acquireDVFSLockLocked : type : DVFS_MIN_LIMIT  frequency : 2457600  uid : 1000  pid : 1055  pkgName : hcm@81
2020-09-21 16:03:25.060 4220-4323/? I/TypeManager: Failed to read /vendor/etc/nnapi_extensions_app_allowlist ; No app allowlisted for vendor extensions use.
2020-09-21 16:03:25.061 4220-4323/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-21 16:03:25.061 4220-4323/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-21 16:03:25.061 4220-4323/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-21 16:03:25.061 4220-4323/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-21 16:03:25.061 747-747/? E/audit: type=1400 audit(1600671805.056:101959): avc:  denied  { read } for  pid=4220 comm="roidJUnitRunner" path="/storage/emulated/0/nnstreamer/test/add/add.tflite" dev="sdcardfs" ino=23663 scontext=u:r:hal_neuralnetworks_default:s0 tcontext=u:object_r:sdcardfs:s0 tclass=file permissive=0 SEPF_SM-G981N_10_0021 audit_filtered
2020-09-21 16:03:25.061 4220-4323/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-21 16:03:25.061 747-747/? E/audit: type=1300 audit(1600671805.056:101959): arch=c00000b7 syscall=29 success=yes exit=0 a0=43 a1=c0306201 a2=7c679f03a0 a3=7c679f06a0 items=0 ppid=765 pid=4220 auid=4294967295 uid=10131 gid=10131 euid=10131 suid=10131 fsuid=10131 egid=10131 sgid=10131 fsgid=10131 tty=(none) ses=4294967295 comm="roidJUnitRunner" exe="/system/bin/app_process64" subj=u:r:untrusted_app_27:s0:c131,c256,c512,c768 key=(null)
2020-09-21 16:03:25.061 4220-4323/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-21 16:03:25.061 747-747/? E/audit: type=1327 audit(1600671805.056:101959): proctitle="org.nnsuite.nnstreamer.test"
2020-09-21 16:03:25.061 4220-4323/? E/nnstreamer.tes: getSupportedOperations_1_2 failure: Status(EX_TRANSACTION_FAILED): 'FAILED_TRANSACTION: '
2020-09-21 16:03:25.062 4220-4323/? E/Manager: IDevice::getSupportedOperations returned the error GENERAL_FAILURE
2020-09-21 16:03:25.062 747-747/? E/audit: type=1400 audit(1600671805.056:101960): avc:  denied  { read } for  pid=4220 comm="roidJUnitRunner" path="/storage/emulated/0/nnstreamer/test/add/add.tflite" dev="sdcardfs" ino=23663 scontext=u:r:hal_neuralnetworks_default:s0 tcontext=u:object_r:sdcardfs:s0 tclass=file permissive=0 SEPF_SM-G981N_10_0021 audit_filtered
2020-09-21 16:03:25.062 747-747/? E/audit: type=1300 audit(1600671805.056:101960): arch=c00000b7 syscall=29 success=yes exit=0 a0=43 a1=c0306201 a2=7c679f03a0 a3=7c679f06a0 items=0 ppid=765 pid=4220 auid=4294967295 uid=10131 gid=10131 euid=10131 suid=10131 fsuid=10131 egid=10131 sgid=10131 fsgid=10131 tty=(none) ses=4294967295 comm="roidJUnitRunner" exe="/system/bin/app_process64" subj=u:r:untrusted_app_27:s0:c131,c256,c512,c768 key=(null)
2020-09-21 16:03:25.062 747-747/? E/audit: type=1327 audit(1600671805.056:101960): proctitle="org.nnsuite.nnstreamer.test"
2020-09-21 16:03:25.062 747-747/? E/audit: type=1400 audit(1600671805.056:101961): avc:  denied  { read } for  pid=4220 comm="roidJUnitRunner" path="/storage/emulated/0/nnstreamer/test/add/add.tflite" dev="sdcardfs" ino=23663 scontext=u:r:hal_neuralnetworks_default:s0 tcontext=u:object_r:sdcardfs:s0 tclass=file permissive=0 SEPF_SM-G981N_10_0021 audit_filtered

...

2020-09-21 16:03:25.093 4220-4323/? E/ExecutionBuilder: Setting argument with invalid length: 8, expected length: 4
2020-09-21 16:03:25.093 4220-4323/? E/tflite: Aborting since NNAPI returned failure nnapi_delegate.cc:805

...

2020-09-21 16:03:25.111 1055-11453/? I/ActivityManager: Process org.nnsuite.nnstreamer.test (pid 4220) has died: fore FGS (2707,4407)
2020-09-21 16:03:25.112 1055-11453/? W/ActivityManager: Crash of app org.nnsuite.nnstreamer.test running instrumentation ComponentInfo{org.nnsuite.nnstreamer.test/android.support.test.runner.AndroidJUnitRunner}
2020-09-21 16:03:25.112 1055-11453/? I/ActivityManager: Force stopping org.nnsuite.nnstreamer.test appid=10131 user=0: finished inst
```

### PR DESCRIPTION
**Pull request by anyj0527**
- Use NNAPI delegate directly instead of deprecated `UseNNAPI` method.
- This resolves #2755.
- Ref: https://github.com/tensorflow/tensorflow/commit/bf6f488c1102a88252960c4c8b90f6122e966b14

Signed-off-by: Yongjoo Ahn <yongjoo1.ahn@samsung.com>

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

TODO: Handle #2820

### CODE REVIEW

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 1]
```diff
@@ -289,12 +298,32 @@ TFLiteInterpreter::loadModel (bool use_nnapi)
...
+#if defined(__ANDROID__) && (TFLITE_VERSION_MAJOR >= 2 || TFLITE_VERSION_MINOR >= 14)
+    stateful_nnapi_delegate.reset (new tflite::StatefulNnApiDelegate ());
+    setDelegate (stateful_nnapi_delegate.get ());
 
-  if (interpreter->AllocateTensors () != kTfLiteOk) {
-    ml_loge ("Failed to allocate tensors\n");
-    return -2;
+    if (interpreter->ModifyGraphWithDelegate (delegate_) != kTfLiteOk) {
+      ml_loge ("Failed to allocate tensors with NNAPI delegate\n");
+          return -2;
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:308]: plz fix indent (line 308, 317  and 323)

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc]: Updated. Thanks!

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc [hunk 2]
```diff
@@ -137,6 +140,12 @@ class TFLiteInterpreter
   int getTensorDim (int tensor_idx, tensor_dim dim);
   int setTensorProp (const std::vector<int> &tensor_idx_list,
       GstTensorsInfo * tensorMeta);
+
+  TfLiteDelegate* delegate_ = nullptr; /**< The delegate for tflite interpreter */
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:144]: Do you really need this pointer? It seems to be same with `stateful_nnapi_delegate.get ()`.

**Comment by jaeyun-jung** [comment]
Discussion (tf-lite gelegate) for future work
- define enum (delegate : nnapi / gpu / ..) in tf-lite sub-plugin and set proper TfLiteDelegate
- change custom option str (UseNNAPI:true > Delegate:NNAPI)

File: ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc (hunk 2 above)
Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc:144]: I declare this pointer for multiple delegate support: NNAPI / gpu / ... .
I'll update this PR reflecting Jaeyun's comment:

> Discussion (tf-lite gelegate) for future work
> 
> * define enum (delegate : nnapi / gpu / ..) in tf-lite sub-plugin and set proper TfLiteDelegate
> * change custom option str (UseNNAPI:true > Delegate:NNAPI)

**Comment by myungjoo** [comment]
> Discussion (tf-lite gelegate) for future work
> 
> * define enum (delegate : nnapi / gpu / ..) in tf-lite sub-plugin and set proper TfLiteDelegate
> * change custom option str (UseNNAPI:true > Delegate:NNAPI)

If you want to decide this quickly, summon all the related (@jaeyun-jung @again4you @myungjoo ) for a 5-min meeting anytime.

**Comment by anyj0527** [comment]
I'll update this PR after #2842

**Review by dongju-chae — APPROVED** [review]
LGTM

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**